### PR TITLE
Fish forces REVIEW NEEDED

### PR DESCRIPTION
**Pull request by reinthal**
Add fish forces to velocity update

### CODE REVIEW

File: code/agents.py [hunk 1]
```diff
@@ -64,13 +65,30 @@ def  __init__(self, position, velocity, fish_id, environment, ann, image = None,
...
+        k = self.environment.k
+        power = self.power
+        for fish in neighbours:
+            print("neihgbour detected at"+str(self.position))
+            f_i = k / np.abs(fish.position-self.position)**power
+            f += f_i
+
+        return f
+
+    def get_neighbours(self):
```
Comment (rasake) [inline code/agents.py]: Good move to have this in dedicated function, improves readability. But change name to find_neighbours, there are no getters in Python.

File: code/agents.py [hunk 2]
```diff
@@ -64,13 +65,30 @@ def  __init__(self, position, velocity, fish_id, environment, ann, image = None,
     self._time_since_last_ANN_update
     self.fish_index
     """
-    
+
+    def get_fish_forces(self):
+        f = np.array([0.0, 0.0])
+        neighbours = self.get_neighbours()
```
Comment (rasake) [inline code/agents.py]: Assume neigbours are already found by think(), just get them directly from the attribute self.neighbouring_fish. That way we don't have to do the search twice.

File: code/agents.py [hunk 3]
```diff
@@ -64,13 +65,30 @@ def  __init__(self, position, velocity, fish_id, environment, ann, image = None,
     self._time_since_last_ANN_update
     self.fish_index
     """
-    
+
+    def get_fish_forces(self):
```
Comment (rasake) [inline code/agents.py]: Don't use get in python method names. Call it something else like "calculate_fish_forces", or make it a property using the @property decorator.

File: code/main.py [hunk 4]
```diff
@@ -78,6 +78,9 @@ def __init__(self, nbr_fish, nbr_predators, boundaries, ann_weights, graphics_on
             fish = Fish(pos, velocity, i, self, ann, fish_image, self.sprite_batch_fishes)
             self.fish_lst.append(fish)
 
+        # Initialize physical constants of  Coloumb force equation
+        self.k = 1
```
Comment (rasake) [inline code/main.py]: Did you intend to put this attributes in each fish instance? Then they must be indented as part of the for loop and it should be fish.k = 1, fish.power=6, in that case I believe.

Comment (reinthal) [reply code/main.py]: No. It's part of the environment. Makes more sense that physics is part of the environment right?

Comment (rasake) [reply code/main.py]: I see.

File: code/agents.py [hunk 5]
```diff
@@ -64,13 +65,32 @@ def  __init__(self, position, velocity, fish_id, environment, ann, image = None,
...
+    def calculate_fish_forces(self):
+        f = np.array([0.0, 0.0])
+        neighbours = self.find_neighbours()
+        k = self.environment.k
+        power = self.environment.power
+        for fish in neighbours:
+
+            k = self.environment.k
+            power = self.environment.power
+            f_i = k / np.abs(fish.position-self.position)**power
```
Comment (rasake) [inline code/agents.py]: f_i should be directed as rel. distanc vector between fish and have magnitude inversely proport. to (norm of rel. distance)^power

**Review by rasake — CHANGES_REQUESTED** [review]

**Comment by kognitiva** [comment]
@rasake I fixed the fish forces now! I set the parameters very high to make sure that it was doing its job. I guess we have to run some optimization before we know what we level of force we need to get nice results.

**Comment by rasake** [comment]
Funkar kraften nu?

**Pull request merged**